mysys/my_uuid_v7.c

```diff
@@ -0,0 +1,139 @@
+/* Copyright (c) 2024, Stefano Petrilli
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; version 2 of the License.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program; if not, write to the Free Software
+   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1335  USA */
+
+/*
+  implements Universal Unique Identifiers version 7, as described in
+  draft-ietf-uuidrev-rfc4122bis-14.
+
+  A UUIDv7 has the following structure:
+
+  Field                     Octet #          Note
+ unix_ts_ms                 0-5     Big-endian unsigned number of
+                                    Unix epoch timestamp in
+                                    milliseconds.
+ version_and_rand_a         6-7     CSPRNG 12 bits multiplexed
+                                    with the version number.
+ variant_and_rand_b         8-15    CSPRNG 62 bits multiplexed
+                                    with the version number.
+*/
+
+#include "mysys_priv.h"
+#include <my_rnd.h>
+#include <m_string.h>
+#include <myisampack.h> /* mi_int2store, mi_int4store */
+
+static my_bool my_uuid_v7_inited= 0;
+static uint64 milliseconds;
+static uint64 uuid_time= 0;
+
+static mysql_mutex_t LOCK_uuid_v7_generator;
+
+#define UUID_VERSION          0x7000
+#define UUID_VERSION_MASK     0x0FFF
+#define UUID_VARIANT          0x8000000000000000
+#define UUID_VARIANT_MASK     0x3FFFFFFFFFFFFFFF
+
+/**
+  Init structures needed for my_uuid_v7
+*/
+void my_uuid_v7_init()
+{
+  if (my_uuid_v7_inited)
+    return;
+  my_uuid_v7_inited= 1;
+  milliseconds= 0;
+
+  mysql_mutex_init(key_LOCK_uuid_v7_generator, &LOCK_uuid_v7_generator, MY_MUTEX_INIT_FAST);
+}
+
+/**
+   Create a global unique identifier version 7(uuidv7)
+
+   @func  my_uuid_v7()
+   @param to   Store uuidv7 here. Must be of size MY_UUID_SIZE (16)
+*/
+void my_uuid_v7(uchar *to)
+{
+  uint64 tv, unix_ts_ms;
+  uint16 rand_a, version_and_rand_a;
+  uint64 rand_b, variant_and_rand_b;
+
+  DBUG_ASSERT(my_uuid_v7_inited);
+
+  tv= my_hrtime().val / 1000;
+
+  // Regulate the access to uuid_time and milliseconds
+  mysql_mutex_lock(&LOCK_uuid_v7_generator);
+
+  if (likely(tv > uuid_time))
+  {
+    /*
+      Current time is ahead of last timestamp, as it should be.
+      If we "borrowed time", give it back, just as long as we
+      stay ahead of the previous timestamp.
```

Comment:
If you build up a large enough "debt" in `milliseconds`, the server process could conceivably be stopped and restarted… and then would be generating "past" `uuid7()` values due to `milliseconds` resetting to 0.

Should there perhaps be a maximum value for `milliseconds` (maybe 1000 ms?) beyond which the `uuid7` function will **block** rather than continue adding to the debt?